intermediate_source/named_tensor_tutorial.py

```diff
@@ -0,0 +1,518 @@
+# `here <https://github.com/facebookresearch/ParlAI/blob/f7db35cba3f3faf6097b3e6b208442cd564783d9/parlai/agents/transformer/modules.py#L907>`_.
+# Read through the code at that example; then, compare with the code below,
+# noting that there are four places labeled (I), (II), (III), and (IV), where
+# using named tensors enables more readable code; we will dive into each of these
+# after the code block.
+
+import torch.nn as nn
+import torch.nn.functional as F
+import math
+
+class MultiHeadAttention(nn.Module):
+    def __init__(self, n_heads, dim, dropout=0):
+        super(MultiHeadAttention, self).__init__()
+        self.n_heads = n_heads
+        self.dim = dim
+
+        self.attn_dropout = nn.Dropout(p=dropout)
+        self.q_lin = nn.Linear(dim, dim)
+        self.k_lin = nn.Linear(dim, dim)
+        self.v_lin = nn.Linear(dim, dim)
+        nn.init.xavier_normal_(self.q_lin.weight)
+        nn.init.xavier_normal_(self.k_lin.weight)
+        nn.init.xavier_normal_(self.v_lin.weight)
+        self.out_lin = nn.Linear(dim, dim)
+        nn.init.xavier_normal_(self.out_lin.weight)
+
+    def forward(self, query, key=None, value=None, mask=None):
+        # (I)
+        query = query.refine_names(..., 'T', 'D')
+        self_attn = key is None and value is None
+        if self_attn:
+            mask = mask.refine_names(..., 'T')
+        else:
+            mask = mask.refine_names(..., 'T', 'T_key')  # enc attn
+
+        dim = query.size('D')
+        assert dim == self.dim, \
+            f'Dimensions do not match: {embedding_dim} query vs {self.dim} configured'
+        assert mask is not None, 'Mask is None, please specify a mask'
+        n_heads = self.n_heads
+        dim_per_head = dim // n_heads
+        scale = math.sqrt(dim_per_head)
+
+        # (II)
+        def prepare_head(tensor):
+            tensor = tensor.refine_names(..., 'T', 'D')
+            return (tensor.unflatten('D', [('H', n_heads), ('D_head', dim_per_head)])
+                          .align_to(..., 'H', 'T', 'D_head'))
+
+        if self_attn:
+            key = value = query
+        elif value is None:
+            # key and value are the same, but query differs
+            key = key.refine_names(..., 'T', 'D')
+            value = key
+        key_len = key.size('T')
+        dim = key.size('D')
+
+        # Distinguish between query_len (T) and key_len (T_key) dims.
+        k = prepare_head(self.k_lin(key)).rename(T='T_key')
+        v = prepare_head(self.v_lin(value)).rename(T='T_key')
+        q = prepare_head(self.q_lin(query))
+
+        dot_prod = q.div_(scale).matmul(k.align_to(..., 'D_head', 'T_key'))
+        dot_prod.refine_names(..., 'H', 'T', 'T_key')  # just a check
+
+        # (III)
+        # Named tensors doesn't support `==` yet; the following is a workaround.
+        attn_mask = (mask.rename(None) == 0).refine_names(*mask.names)
+        attn_mask = attn_mask.align_as(dot_prod)

```

Comment:
We actually do support == now! Yay! (It got into the release branch)

Could we change
```
# Named tensors doesn't support `==` yet; the following is a workaround.
attn_mask = (mask.rename(None) == 0).refine_names(*mask.names)
attn_mask = attn_mask.align_as(dot_prod)
```

to: 
```
attn_mask = (attn_mask == 0).align_as(dot_prod)
```

Reply:
(And also change the code in III)